src/core/socialprovider.js

```diff
@@ -477,12 +478,64 @@ XMPPSocialProvider.prototype.onOnline = function(continuation) {
   this.vCardStore.updateProperty(this.id, 'status', 'ONLINE');
   this.vCardStore.refreshContact(this.id, null);
 
+  this.pollForDisconnect_();
+
   continuation(this.vCardStore.getClient(this.id));
 };
 
+XMPPSocialProvider.prototype.ping_ = function() {
+  var pingDate = new Date();
+  var ping = new window.XMPP.Element('iq', {type: 'get'})
+      .c('ping', {'xmlns': 'urn:xmpp:ping'}).up();
+  this.client.send(ping);
+
+  // Check that we got a response from the server after the ping was sent.
+  setTimeout(function() {
+    if (this.client &&  // check this.client to be sure logout wasn't called.
+        (!this.lastMessageDate_ || this.lastMessageDate_ < pingDate)) {
+      // No response to ping, we are disconnected.
+      console.warn('No ping response from server, logging out');
+      this.logout();
+    }
+  }.bind(this), 10000);
```

Comment:
BTW the 2 seconds that you see is related to detecting for sleep.  The way that works is we have a callback that is supposed to run every second.  In that callback we check when the last time the callback was run, and if that is 2 seconds or more (when it should be only 1 second), this means the computer was likely asleep (because our JS code stopped running).